### DISCUSSION TITLE
Custom Function for Bot composer

### DISCUSSION THREAD

**Discussion by bhargavikalicheti** [microsoft/BotFramework-Composer#9086] [Design]
Is there a way to create a custom function for bot in cases where we could n't use prebuilt functions?

**Comment by guitarguySD** [comment]
Yes,

google bot framework composer custom function example

**Reply by MI14251** [reply]
you are great..

**Comment by OEvgeny** [comment]
Here is the link: https://learn.microsoft.com/en-us/composer/how-to-create-custom-actions

Closing this. Feel free to reopen

**Reply by MI14251** [reply]
I need to call a C# class library method from .lg file by using the customFunctions [namespace.staticclass.methodname] attribute in the appsettings. Do you have any sample on Custom Functions (C# code) that can be used in .lg file in Bot Framework Composer v2.1.2?

**Reply by OEvgeny** [reply]
I haven't seen something similar. The only docs I found are:

https://learn.microsoft.com/en-us/azure/bot-service/language-generation/functions-injected-from-language-generation?view=azure-bot-service-4.0

Which mention injecting functions, but I haven't found where it is covered

**Comment by mrw137** [comment]
In your `Startup.cs` file add your function as an adaptive expression function. Don't forget to use [`AdaptiveExpressions`](https://learn.microsoft.com/en-us/dotnet/api/adaptiveexpressions.expression.functions?view=botbuilder-dotnet-stable#adaptiveexpressions-expression-functions).

```C#
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.StaticFiles;
using Microsoft.Bot.Builder.Dialogs.Adaptive.Runtime.Extensions;
using Microsoft.Bot.Builder.Integration.AspNet.Core;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using AdaptiveExpressions; // Add this

namespace Empty
{
    public class Startup
    {
        public Startup(IConfiguration configuration)
        {
            Configuration = configuration;
        }

        public IConfiguration Configuration { get; }

        // This method gets called by the runtime. Use this method to add services to the container.
        public void ConfigureServices(IServiceCollection services)
        {
            services.AddControllers().AddNewtonsoftJson(options =>
            {
                options.SerializerSettings.MaxDepth = HttpHelper.BotMessageSerializerSettings.MaxDepth;
            });
            services.AddBotRuntime(Configuration);
        }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            Expression.Functions.Add("myCustomFunction", (args) =>
            {
                return "customFunctionReturnValue";
            });
.
.
.
```

Then in your `appsettings.json`, add the name of your function in `customFunctions`, wrap it in astrix so Bot Framework Composer doesn't complain.

```json
{
  "customFunctions": [
    "*myCustomFunction*"
  ],
.
.
.
```

Then use it in your `.lg` file like an other adaptive expression.
```
[import](common.lg)

# SendActivity_myCustomFunction()
[Activity
    Text = ${SendActivity_myCustomFunction_text()}
]

# SendActivity_myCustomFunction_text()
- ${myCustomFunction()}
```
